jsonpb/jsonpb.go

```diff
@@ -73,8 +73,40 @@ type Marshaler struct {
 
 	// Whether to use the original (.proto) name for fields.
 	OrigName bool
+
+	// A custom URL resolver to use when marshaling Any messages to JSON.
+	// If unset, DefaultUrlResolver is used.
+	UrlResolver UrlResolver
+}
+
+// UrlResolver resolves a type URL into an instance of the associated message.
+// This is used for serializing and de-serializing Any messages to and from
+// JSON.
+type UrlResolver interface {
+	Resolve(typeUrl string) (proto.Message, error)
+}
+
+type defaultResolver struct{}
+
+func (defaultResolver) Resolve(typeUrl string) (proto.Message, error) {
+	// Only the part of typeUrl after the last slash is relevant.
+	mname := typeUrl
+	if slash := strings.LastIndex(mname, "/"); slash >= 0 {
+		mname = mname[slash+1:]
+	}
+	mt := proto.MessageType(mname)
+	if mt == nil {
+		return nil, fmt.Errorf("unknown message type %q", mname)
+	}
+	return reflect.New(mt.Elem()).Interface().(proto.Message), nil
 }
 
+// DefaultUrlResolver can resolve any generated protobuf message that is
+// registered via proto.RegisterType. The fully-qualified message name is first
+// extracted from the type URL (e.g. the last path element). The resolver then
+// invokes proto.MessageType(string) to create a message instance for that name.
+var DefaultUrlResolver UrlResolver = defaultResolver{}
```

Comment:
I don't think there is a need to export an instance of this.  If a custom resolver wants to use the global registry, then it can do so on its own.

Reply:
Removed.